security/experimental_authenticators.rst

```diff
@@ -0,0 +1,432 @@
+            'firewalls' => [
+                'main' => [
+                    // allow authentication using a form or HTTP basic
+                    'form_login' => null,
+                    'http_basic' => null,
+
+                    // configure the form authentication as the entry point for unauthenticated users
+                    'entry_point' => 'form_login'
+                ],
+            ],
+        ]);
+
+.. note::
+
+    You can also create your own authentication entry point by creating a
+    class that implements
+    :class:`Symfony\\Component\\Security\\Http\\EntryPoint\\AuthenticationEntryPointInterface`.
+    You can then set ``entry_point`` to the FQCN (e.g.
+    ``entry_point: App\Security\CustomEntryPoint``)
+
+.. _authenticators-removed-authentication-providers:
+
+Creating a Custom Authenticator
+-------------------------------
+
+Security traditionally could be extended by writing
+:doc:`custom authentication providers </security/custom_authentication_provider>`.
+The authenticator-based system dropped support for these providers and
+introduced a new authenticator interface as a base for custom
+authentication methods.
+
+.. tip::
+
+    :doc:`Guard authenticators </security/guard_authentication>` are still
+    supported in the authenticator-based system. It is however recommended
+    to also update these when you're refactoring your application to the
+    new system.
+
+    The new authenticator interface has quite some overlap with the guard
+    authenticator interface, making the rewrite easier. To avoid duplicated
+    class names, it is recommended to remove the ``symfony/security-guard``
+    package before rewriting your Security classes:
+
+    .. code-block:: terminal
+
+        $ composer remove symfony/security-guard
+
+Authenticators should implement the
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.
+The easiest way to use it is by extending
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AbstractAuthenticator`::
+
+    // src/Security/ApiKeyAuthenticator.php
+    namespace App\Security;
+
+    use App\Entity\User;
+    use Doctrine\ORM\EntityManagerInterface;
+    use Symfony\Component\HttpFoundation\JsonResponse;
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+    use Symfony\Component\Security\Core\Exception\AuthenticationException;
+    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+    use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
+    use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+
+    class ApiKeyAuthenticator extends AbstractAuthenticator
```

Comment:
I'm not sure I like having an API token authenticator, as it doesn't use the default `Passport` and has no success response. Doing a login form isn't great either (users should extend `AbstractFormLoginAuthenticator` then).
Does anyone have an idea for a good example authenticator that we can use?

Reply:
What about a magic link auth?
Retrieves and checks data from request (ie a user uniq token)
If ok, a response that redirects to a default authenticated page

Reply:
I really liked this idea, until I realized this also used the `SelfValidatingPassport`. In the end, I think we can keep the current example for now. Let's see if it's confusing.